src/tqec/compile/specs/library/base.py

```diff
@@ -0,0 +1,449 @@
+from tqec.templates.base import RectangularTemplate
+from tqec.templates.enums import ZObservableOrientation
+from tqec.utils.enums import Basis
+from tqec.utils.exceptions import TQECException
+from tqec.utils.position import Direction3D
+from tqec.utils.scale import LinearFunction
+
+
+class BaseCubeBuilder(CubeBuilder):
+    """Base implementation of the :class:`~tqec.compile.specs.base.CubeBuilder`
+    interface.
+
+    This class provides a good enough default implementation that should be
+    enough for most of the cube builders.
+    """
+
+    DEFAULT_BLOCK_REPETITIONS: Final[LinearFunction] = LinearFunction(2, -1)
+
+    def __init__(self, compiler: PlaquetteCompiler) -> None:
+        """Initialise the :class:`BaseCubeBuilder` with a compiler.
+
+        Args:
+            compiler: compiler to transform the plaquettes in the standard
+                implementation to a custom implementation.
+        """
+        self._translator = DefaultRPNGTranslator()
+        self._compiler = compiler
+
+    def _get_template_and_plaquettes(
+        self,
+        spec: CubeSpec,
+    ) -> tuple[
+        RectangularTemplate,
+        tuple[Plaquettes, Plaquettes, Plaquettes],
+    ]:
+        """Get the template and plaquettes corresponding to the provided ``spec``.
+
+        Args:
+            spec: specification of the cube we want to implement.
+
+        Returns:
+            the template and list of 3 mappings from plaquette indices to RPNG
+            descriptions that are needed to implement the cube corresponding to
+            the provided ``spec``.
+        """
+        assert isinstance(spec.kind, ZXCube)
+        x, _, z = spec.kind.as_tuple()
+        if not spec.is_spatial:
+            orientation = (
+                ZObservableOrientation.HORIZONTAL
+                if x == Basis.Z
+                else ZObservableOrientation.VERTICAL
+            )
+
+            return get_memory_qubit_raw_template(), (
+                Plaquettes(
+                    get_memory_qubit_rpng_descriptions(orientation, z, None).map_values(
+                        self._get_plaquette
+                    )
+                ),
+                Plaquettes(
+                    get_memory_qubit_rpng_descriptions(
+                        orientation, None, None
+                    ).map_values(self._get_plaquette)
+                ),
+                Plaquettes(
+                    get_memory_qubit_rpng_descriptions(orientation, None, z).map_values(
+                        self._get_plaquette
+                    )
+                ),
+            )
+        # else:
+        return get_spatial_cube_qubit_raw_template(), (
+            get_spatial_cube_qubit_plaquettes(x, spec.spatial_arms, z, None),
+            get_spatial_cube_qubit_plaquettes(x, spec.spatial_arms, None, None),
+            get_spatial_cube_qubit_plaquettes(x, spec.spatial_arms, None, z),
+        )
```

Comment:
Even though spatial junctions will be adressed in a follow-up PR, I think it would be nice to use the `*_rpng_descriptions` functions here too.